### PR TITLE
feat: Survey 10146 input auto focus

### PR DESCRIPTION
**Pull request by flying3615**
Auto focus on TextInput

Author Checklist

- [ ] appropriate description or links provided to provide context on the PR
- [ ] self reviewed, seems easy to understand and follow
- [ ] reasonable code test coverage
- [ ] change is documented in Storybook and/or markdown files

Reviewer Checklist

- Follows convention
- Does what the author says it will do
- Does not appear to cause side effects and breaking changes
  - if it does cause breaking changes, those are appropriately referenced

Post merge

- [ ] Post about the change in #lui-cop

Conventional Commit Cheat Sheet:
**build**: Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
**ci**: Changes to our CI configuration files and scripts (example scopes: Circle, BrowserStack, SauceLabs)
**docs**: Documentation only changes
**feat**: A new feature
**fix**: A bug fix
**perf**: A code change that improves performance
**refactor**: A code change that neither fixes a bug nor adds a feature
**test**: Adding missing tests or correcting existing tests

### CODE REVIEW

File: src/lui/TextInputFormatted.tsx [hunk 1]
```diff
@@ -18,12 +18,19 @@ export interface LuiTextInputProps extends DetailedHTMLProps<InputHTMLAttributes
 }
 
 export const TextInputFormatted = (props: LuiTextInputProps): ReactElement => {
+  const inputReference = useRef<HTMLInputElement>(null);
+  useEffect(() => {
+    if (inputReference && inputReference.current) {
+      inputReference.current.focus();
```
Comment (matttdawson) [inline src/lui/TextInputFormatted.tsx:24]: inputReference.current?.focus();

Comment (matttdawson) [reply src/lui/TextInputFormatted.tsx:24]: Does this effect any other components that use TextInputFormatted?

Comment (matttdawson) [reply src/lui/TextInputFormatted.tsx:24]: This seems to fight with react menu attempting to select the first menu item on popup.
It feels like a race condition

Comment (flying3615) [reply src/lui/TextInputFormatted.tsx]: Rewrite the logic to click the MenuRadioItem, then find the first input element in MenuSubComponent to focus on

File: src/components/gridForm/GridFormMultiSelect.tsx [hunk 2]
```diff
@@ -191,11 +193,23 @@ export const GridFormMultiSelect = <RowType extends GridBaseRow>(props: GridForm
...
+                            <MenuRadioItem
+                              item={item}
+                              options={options}
+                              setOptions={setOptions}
+                              onClick={() => {
+                                // Default to focus on first input in subComponent
+                                defer(() => {
+                                  if (firstInputSubComponent.current) {
+                                    (firstInputSubComponent.current as HTMLInputElement).focus();
+                                  }
```
Comment (matttdawson) [inline src/components/gridForm/GridFormMultiSelect.tsx:205]: firstInputSubComponent.current?.focus()

File: src/components/gridForm/GridFormMultiSelect.tsx [hunk 3]
```diff
@@ -395,17 +411,46 @@ const MenuRadioItem = (props: {
...
+    }
+
+    return null;
+  }, []);
+
+  useEffect(() => {
+    if (focusableRef.current) {
+      const inputElement = findFirstChildInputElement(focusableRef.current);
+      if (inputElement) {
+        (ref as React.MutableRefObject<HTMLInputElement>).current = inputElement;
```
Comment (matttdawson) [inline src/components/gridForm/GridFormMultiSelect.tsx:448]: Why do you need to do this?

File: src/components/gridForm/GridFormMultiSelect.tsx [hunk 4]
```diff
@@ -369,6 +384,7 @@ const MenuRadioItem = (props: {
           e.keepOpen = true;
           toggleValue(item);
         }
+        props.onClick && props.onClick();
```
Comment (matttdawson) [inline src/components/gridForm/GridFormMultiSelect.tsx:387]: This should only invoke if the checkbox enables

Comment (matttdawson) [reply src/components/gridForm/GridFormMultiSelect.tsx:387]: Maybe call it onChecked ?

File: src/components/gridForm/GridFormMultiSelect.tsx [hunk 5]
```diff
@@ -395,17 +411,46 @@ const MenuRadioItem = (props: {
...
+    setOptions: (options: MultiSelectOption[]) => void;
+    subComponentIsValid: Record<string, boolean>;
+    triggerSave: () => Promise<void>;
+  },
+  ref: ForwardedRef<HTMLInputElement>,
+) {
   const { data, item, options, setOptions, subComponentIsValid, triggerSave } = props;
+  const focusableRef = React.useRef<HTMLElement | null>(null);
+  const findFirstChildInputElement = useCallback((element: HTMLElement): HTMLInputElement | null => {
+    if (element.tagName === "INPUT") {
```
Comment (matttdawson) [inline src/components/gridForm/GridFormMultiSelect.tsx:430]: Could all this just be a:
```
element.querySelectorAll("input")[0]
```

File: src/components/gridForm/GridFormMultiSelect.tsx (hunk 3 above)
Comment (flying3615) [reply src/components/gridForm/GridFormMultiSelect.tsx:448]: Here just to associate the first input with ref, so in the parent component ref can be handled for focusing

Comment (matttdawson) [reply src/components/gridForm/GridFormMultiSelect.tsx:436]: Why do you need the "as" I mean

Comment (matttdawson) [reply src/components/gridForm/GridFormMultiSelect.tsx:436]: ref.current = inputElement ?? null;

File: src/components/gridForm/GridFormMultiSelect.tsx [hunk 6]
```diff
@@ -369,6 +382,7 @@ const MenuRadioItem = (props: {
           e.keepOpen = true;
           toggleValue(item);
         }
+        props.onChecked && props.onChecked();
```
Comment (matttdawson) [inline src/components/gridForm/GridFormMultiSelect.tsx:385]: item.checked && props.onChecked && props.onChecked();

File: src/components/gridForm/GridFormMultiSelect.tsx (hunk 3 above)
Comment (matttdawson) [reply src/components/gridForm/GridFormMultiSelect.tsx:436]: ref.current = element.querySelectorAll("input")[0] ?? null;

Comment (flying3615) [reply src/components/gridForm/GridFormMultiSelect.tsx]: `ForwardedRef` need to check if a function or a MutableRefObject to avoid the type casting

**Review by matttdawson — APPROVED** [review]
LGTM

**Pull request merged**